### PR TITLE
refactor: encryption refactor

### LINKED ISSUE #2346: Encryption provider
Ah I see what you were saying now - that it would have been preferable to inject the fully constructed encryptor. 

I think the way to decouple this would be to have a "subkey provider" which the DAL implements. Then you can pass in an encryptor _builder_, the DAL calls `builder.Build(subkeyProvider)` to construct the encryptor.

LMK if that needs clarification.

_Originally posted by @alecthomas in https://github.com/TBD54566975/ftl/pull/2338#discussion_r1715981630_

### PR DESCRIPTION
**Pull request by gak**
Fixes #2346

### CODE REVIEW

File: backend/controller/cronjobs/cronjobs_test.go [hunk 1]
```diff
@@ -37,7 +38,7 @@ func TestServiceWithMockDal(t *testing.T) {
 		attemptCountMap: map[string]int{},
 	}
 	conn := sqltest.OpenForTesting(ctx, t)
-	parentDAL, err := db.New(ctx, conn, optional.None[string]())
+	parentDAL, err := db.New(ctx, conn, *encryption.NewBuilder())
```
Comment (alecthomas) [inline backend/controller/cronjobs/cronjobs_test.go:41]: It's not really idiomatic to deref pointers as part of normal code flow, can this and above just return a value instead?

File: backend/controller/dal/async_calls.go [hunk 2]
```diff
@@ -115,7 +115,7 @@ func (d *DAL) AcquireAsyncCall(ctx context.Context) (call *AsyncCall, err error)
 		return nil, fmt.Errorf("failed to parse origin key %q: %w", row.Origin, err)
 	}
 
-	decryptedRequest, err := d.decrypt(encryption.AsyncSubKey, row.Request)
+	decryptedRequest, err := d.decrypt(encryption.AsyncSubkey{}, row.Request)
```
Comment (alecthomas) [inline backend/controller/dal/async_calls.go:118]: Unfortunately this is going to conflict with https://github.com/TBD54566975/ftl/pull/2400 once that's merged.

File: backend/controller/dal/encryption.go [hunk 3]
```diff
@@ -58,49 +58,31 @@ func (d *DAL) decryptJSON(subKey encryption.SubKey, encrypted []byte, v any) err
...
 
-	url, ok := d.kmsURL.Get()
-	if !ok {
-		logger.Infof("KMS URL not set, encryption not enabled")
-		d.encryptor = encryption.NewNoOpEncryptor()
-		return nil
-	}
+	encryptedKey, err = tx.db.GetOnlyEncryptionKey(ctx)
+	if err != nil && dal.IsNotFound(err) {
+		logger.Infof("No encryption key found, generating a new one")
```
Comment (alecthomas) [inline backend/controller/dal/encryption.go:71]: Let's make this debug. In general we're putting most logs at debug, so they don't show up for the user during development mode.

File: internal/encryption/encryption.go [hunk 4]
```diff
@@ -16,30 +18,91 @@ import (
...
+	b.kmsURI = optional.Ptr[string](kmsURI)
+	return b
+}
+
+func (b *Builder) WithAWSV1Client(kms *awsv1kms.KMS) *Builder {
+	b.awsV1Client = kms
+	return b
+}
+
+func (b *Builder) Build(ctx context.Context, provider KeyStoreProvider) (DataEncryptor, error) {
```
Comment (alecthomas) [inline internal/encryption/encryption.go:68]: Nice

File: internal/encryption/encryption.go [hunk 5]
```diff
@@ -16,30 +18,91 @@ import (
...
+}
+
+// WithKMSURI sets the URI for the KMS key to use. Omitting this call or using nil will create a NoOpEncryptor.
+func (b *Builder) WithKMSURI(kmsURI *string) *Builder {
+	b.kmsURI = optional.Ptr[string](kmsURI)
+	return b
+}
+
+func (b *Builder) WithAWSV1Client(kms *awsv1kms.KMS) *Builder {
+	b.awsV1Client = kms
```
Comment (alecthomas) [inline internal/encryption/encryption.go:64]: This doesn't seem to be used, can it be deleted?

File: internal/encryption/encryption.go [hunk 6]
```diff
@@ -16,30 +18,91 @@ import (
...
+}
+
+func NewBuilder() *Builder {
+	return &Builder{
+		kmsURI: optional.None[string](),
+	}
+}
+
+// WithKMSURI sets the URI for the KMS key to use. Omitting this call or using nil will create a NoOpEncryptor.
+func (b *Builder) WithKMSURI(kmsURI *string) *Builder {
```
Comment (alecthomas) [inline internal/encryption/encryption.go:58]: For builders in Go it's better for them to be "immutable" (or as close as Go gets), by making the receiver a value.

Also if `kmsURI` is intended to be optional it should just be an `Optional` here, the fact that it's a pointer from the CLI struct shouldn't leak down to this abstraction layer.

**Review by alecthomas — COMMENTED** [review]
Nice! Unfortunately this is going to conflict with https://github.com/TBD54566975/ftl/pull/2400 from Friday, but not much we can do about it now 🙁

File: backend/controller/cronjobs/cronjobs_test.go (hunk 1 above)
Comment (alecthomas) [reply backend/controller/cronjobs/cronjobs_test.go:41]: Looking down further, the `Builder` should just be a value everywhere, so I think these pointers can be removed.

File: backend/controller/dal/async_calls.go (hunk 2 above)
Comment (gak) [reply backend/controller/dal/async_calls.go:118]: I will revert those lines!

File: internal/encryption/encryption.go (hunk 5 above)
Comment (gak) [reply internal/encryption/encryption.go:64]: Good catch. I thought I needed it for one of the tests but turns out it got bypassed. YAGNI.

**Review by alecthomas — APPROVED** [review]

**Comment by alecthomas** [comment]
Awesome, this feels pretty clean!

**Pull request merged**